<center>
<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-RP0321EN-SkillsNetwork/labs/module_1/images/SN_web_lightmode.png" width="300"> 
</center>


<h1>Data Wrangling with dplyr</h1>

Estimated time needed: **60** minutes


## Lab Overview:

In this lab, you will focus on wrangling the Seoul bike-sharing demand historical dataset. This is the core dataset to build a predictive model later.

It contains the following columns:

- `DATE` : Year-month-day
- `RENTED BIKE COUNT`- Count of bikes rented at each hour
- `HOUR`- Hour of he day
- `TEMPERATURE` - Temperature in Celsius
- `HUMIDITY` - Unit is `%`
- `WINDSPEED` - Unit is `m/s`
- `VISIBILITY` - Multiplied by 10m
- `DEW POINT TEMERATURE` - The temperature to which the air would have to cool down in order to reach saturation, unit is Celsius
- `SOLAR RADIATION` - MJ/m2
- `RAINFALL` - mm
- `SNOWFALL` - cm
- `SEASONS` - Winter, Spring, Summer, Autumn
- `HOLIDAY` - Holiday/No holiday
- `FUNCTIONAL DAY` - NoFunc(Non Functional Hours), Fun(Functional hours)


For this dataset, you will be asked to use `tidyverse` to perform the following data wrangling tasks:

- **TASK: Detect and handle missing values**
- **TASK: Create indicator (dummy) variables for categorical variables**
- **TASK: Normalize data**

Let's start!


First import the necessary library for this data wrangling task:


In [1]:
# Check if you need to install the `tidyverse` library
require("tidyverse")
library(tidyverse)

Loading required package: tidyverse

── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──

✔ ggplot2 3.3.6     ✔ purrr   0.3.4
✔ tibble  3.1.7     ✔ dplyr   1.0.9
✔ tidyr   1.2.0     ✔ stringr 1.4.0
✔ readr   2.1.2     ✔ forcats 0.5.1

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()



Then load the bike-sharing system data from the csv processed in the previous lab:


In [3]:
bike_sharing_df <- read_csv("raw_seoul_bike_sharing.csv")

Rows: 8760 Columns: 14
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (4): DATE, SEASONS, HOLIDAY, FUNCTIONING_DAY
dbl (10): RENTED_BIKE_COUNT, HOUR, TEMPERATURE, HUMIDITY, WIND_SPEED, VISIBI...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


In [ ]:
# Or you may read it from here again
# url <- "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-RP0321EN-SkillsNetwork/labs/datasets/raw_seoul_bike_sharing.csv"
# Notice some column names in the raw datasets are not standalized if you haven't done them properly in the previous lab

First take a quick look at the dataset:


In [4]:
summary(bike_sharing_df)
dim(bike_sharing_df)

     DATE           RENTED_BIKE_COUNT      HOUR        TEMPERATURE    
 Length:8760        Min.   :   2.0    Min.   : 0.00   Min.   :-17.80  
 Class :character   1st Qu.: 214.0    1st Qu.: 5.75   1st Qu.:  3.40  
 Mode  :character   Median : 542.0    Median :11.50   Median : 13.70  
                    Mean   : 729.2    Mean   :11.50   Mean   : 12.87  
                    3rd Qu.:1084.0    3rd Qu.:17.25   3rd Qu.: 22.50  
                    Max.   :3556.0    Max.   :23.00   Max.   : 39.40  
                    NA's   :295                       NA's   :11      
    HUMIDITY       WIND_SPEED      VISIBILITY   DEW_POINT_TEMPERATURE
 Min.   : 0.00   Min.   :0.000   Min.   :  27   Min.   :-30.600      
 1st Qu.:42.00   1st Qu.:0.900   1st Qu.: 940   1st Qu.: -4.700      
 Median :57.00   Median :1.500   Median :1698   Median :  5.100      
 Mean   :58.23   Mean   :1.725   Mean   :1437   Mean   :  4.074      
 3rd Qu.:74.00   3rd Qu.:2.300   3rd Qu.:2000   3rd Qu.: 14.800      
 Max.   :98.

[1] 8760   14

From the summary, we can observe that:

Columns `RENTED_BIKE_COUNT`, `TEMPERATURE`, `HUMIDITY`, `WIND_SPEED`, `VISIBILITY`, `DEW_POINT_TEMPERATURE`, `SOLAR_RADIATION`, `RAINFALL`, `SNOWFALL` are numerical variables/columns and require normalization. Moreover, `RENTED_BIKE_COUNT` and `TEMPERATURE` have some missing values (NA's) that need to be handled properly.

`SEASONS`, `HOLIDAY`, `FUNCTIONING_DAY` are categorical variables which need to be converted into indicator columns or dummy variables.
Also, `HOUR` is read as a numerical variable but it is in fact a categorical variable with levels ranging from 0 to 23.

Now that you have some basic ideas about how to process this bike-sharing demand dataset, let's start working on it!


# TASK: Detect and handle missing values


 The `RENTED_BIKE_COUNT` column has about 295 missing values, and `TEMPERATURE` has about 11 missing values. Those missing values could be caused by not being recorded, or from malfunctioning bike-sharing systems or weather sensor networks. In any cases, the identified missing values have to be properly handled.


Let's first handle missing values in `RENTED_BIKE_COUNT` column:


Considering `RENTED_BIKE_COUNT` is the response variable/dependent variable, i.e., we want to predict the `RENTED_BIKE_COUNT` using other predictor/independent variables later, and we normally can not allow missing values for the response variable, so missing values for response variable must be either dropped or imputed properly. 

We can see that `RENTED_BIKE_COUNT` only has about 3% missing values (295 / 8760). As such, you can safely drop any rows whose `RENTED_BIKE_COUNT` has missing values.


_TODO:_ Drop rows with missing values in the `RENTED_BIKE_COUNT` column


In [5]:
# Drop rows with `RENTED_BIKE_COUNT` column == NA
seoul_bike_sharing <- bike_sharing_df %>% filter(!is.na(RENTED_BIKE_COUNT))

In [6]:
# Print the dataset dimension again after those rows are dropped
dim(seoul_bike_sharing)

[1] 8465   14

Now that you have handled  missing values in the `RENTED_BIKE_COUNT` variable, let's continue processing missing values for the `TEMPERATURE` column.


Unlike the `RENTED_BIKE_COUNT` variable, `TEMPERATURE` is not a response variable. However, it is still an important predictor variable - as you could imagine, there may be a positve correlation between `TEMPERATURE` and `RENTED_BIKE_COUNT`. For example, in winter time with lower temperatures, people may not want to ride a bike, while in summer with nicer weather, they are more likely to rent a bike.


How do we handle missing values for `TEMPERATURE`? We could simply remove the rows but it's better to impute them because `TEMPERATURE` should be relatively easy and reliable to estimate statistically.


Let's first take a look at the missing values in the TEMPERATURE column.


In [7]:
bike_sharing_df %>% 
                filter(is.na(TEMPERATURE))

DATE       RENTED_BIKE_COUNT HOUR TEMPERATURE HUMIDITY WIND_SPEED VISIBILITY
1  07/06/2018 3221              18   NA          57       2.7        1217      
2  12/06/2018 1246              14   NA          45       2.2        1961      
3  13/06/2018 2664              17   NA          57       3.3         919      
4  17/06/2018 2330              17   NA          58       3.3         865      
5  20/06/2018 2741              19   NA          61       2.7        1236      
6  30/06/2018 1144              13   NA          87       1.7         390      
7  05/07/2018  827              10   NA          75       1.1        1028      
8  11/07/2018  634               9   NA          96       0.6         450      
9  12/07/2018  593               6   NA          93       1.1         852      
10 21/07/2018  347               4   NA          77       1.2        1203      
11 21/08/2018 1277              23   NA          75       0.1        1892      
   DEW_POINT_TEMPERATURE SOLAR_RADIATION RAINFALL SNOWFALL SEASONS HOLIDAY   
1  16.4                  0.96            0.0      0        Summer  No Holiday
2  12.7                  1.39            0.0      0        Summer  No Holiday
3  16.4                  0.87            0.0      0        Summer  No Holiday
4  16.7                  0.66            0.0      0        Summer  No Holiday
5  17.5                  0.60            0.0      0        Summer  No Holiday
6  23.2                  0.71            3.5      0        Summer  No Holiday
7  20.8                  1.22            0.0      0        Summer  No Holiday
8  24.9                  0.41            0.0      0        Summer  No Holiday
9  24.3                  0.01            0.0      0        Summer  No Holiday
10 21.2                  0.00            0.0      0        Summer  No Holiday
11 20.8                  0.00            0.0      0        Summer  No Holiday
   FUNCTIONING_DAY
1  Yes            
2  Yes            
3  Yes            
4  Yes            
5  Yes            
6  Yes            
7  Yes            
8  Yes            
9  Yes            
10 Yes            
11 Yes

It seems that all of the missing values for `TEMPERATURE` are found in `SEASONS == Summer`, so it is reasonable to impute those missing values with the summer average temperature.


_TODO:_ Impute missing values for the TEMPERATURE column using its mean value.


In [8]:
# Calculate the summer average temperature
summer_avg_temp <- seoul_bike_sharing %>%
  filter(SEASONS == "Summer") %>%
  select(TEMPERATURE) %>%
  summarise(mean(TEMPERATURE, na.rm = TRUE)) %>%
  unlist() %>%
  unname()
summer_avg_temp 

[1] 26.58771

In [9]:
# Impute missing values for TEMPERATURE column with summer average temperature
seoul_bike_sharing$TEMPERATURE <- replace_na(seoul_bike_sharing$TEMPERATURE, summer_avg_temp)

In [10]:
# Print the summary of the dataset again to make sure no missing values in all columns
seoul_bike_sharing %>%
  filter(is.na(TEMPERATURE))

DATE RENTED_BIKE_COUNT HOUR TEMPERATURE HUMIDITY WIND_SPEED VISIBILITY
     DEW_POINT_TEMPERATURE SOLAR_RADIATION RAINFALL SNOWFALL SEASONS HOLIDAY
     FUNCTIONING_DAY

In [11]:
# Save the dataset as `seoul_bike_sharing.csv`
write.csv(seoul_bike_sharing, "seoul_bike_sharing.csv")

# TASK: Create indicator (dummy) variables for categorical variables


Regression models can not process categorical variables directly, thus we need to convert them into indicator variables. 


In the bike-sharing demand dataset, `SEASONS`, `HOLIDAY`, `FUNCTIONING_DAY` are categorical variables. 
Also, `HOUR` is read as a numerical variable but it is in fact a categorical variable with levels ranged from 0 to 23.


_TODO:_ Convert `HOUR` column from numeric into character first:


In [12]:
# Using mutate() function to convert HOUR column into character type
seoul_bike_sharing <- seoul_bike_sharing %>% mutate(HOUR = as.character(HOUR))

`SEASONS`, `HOLIDAY`, `FUNCTIONING_DAY`,  `HOUR` are all character columns now and are ready to be converted into indicator variables.

For example, `SEASONS` has four categorical values: `Spring`, `Summer`, `Autumn`, `Winter`. We thus need to create four indicator/dummy variables `Spring`, `Summer`, `Autumn`, and `Winter` which only have the value 0 or 1.

So, given a data entry with the value `Spring` in the `SEASONS` column, the values for the four new columns `Spring`, `Summer`, `Autumn`, and `Winter` will be set to 1 for `Spring` and 0 for the others:

|Spring|Summer|Autumn|Winter|
|----- |------|------|------|
|     1|     0|     0|     0| 


_TODO:_ Convert `SEASONS`, `HOLIDAY`, `FUNCTIONING_DAY`, and `HOUR` columns into indicator columns.

Note that if `FUNCTIONING_DAY` only contains one categorical value after missing values removal, then you don't need to convert it to an indicator column.


In [13]:
# Convert SEASONS, HOLIDAY, FUNCTIONING_DAY, and HOUR columns into indicator columns.
col <- c("SEASONS", "HOLIDAY","FUNCTIONING_DAY", "HOUR")
feature <- function(x) {
  for (x in col) {
    seoul_bike_sharing <<- seoul_bike_sharing %>%
      mutate(dummy = 1) %>%
      spread(key = x, value = dummy, fill = 0)
  }
}
feature()

In [14]:
# Print the dataset summary again to make sure the indicator columns are created properly
summary(seoul_bike_sharing)

     DATE           RENTED_BIKE_COUNT  TEMPERATURE        HUMIDITY    
 Length:8465        Min.   :   2.0    Min.   :-17.80   Min.   : 0.00  
 Class :character   1st Qu.: 214.0    1st Qu.:  3.00   1st Qu.:42.00  
 Mode  :character   Median : 542.0    Median : 13.50   Median :57.00  
                    Mean   : 729.2    Mean   : 12.77   Mean   :58.15  
                    3rd Qu.:1084.0    3rd Qu.: 22.70   3rd Qu.:74.00  
                    Max.   :3556.0    Max.   : 39.40   Max.   :98.00  
   WIND_SPEED      VISIBILITY   DEW_POINT_TEMPERATURE SOLAR_RADIATION 
 Min.   :0.000   Min.   :  27   Min.   :-30.600       Min.   :0.0000  
 1st Qu.:0.900   1st Qu.: 935   1st Qu.: -5.100       1st Qu.:0.0000  
 Median :1.500   Median :1690   Median :  4.700       Median :0.0100  
 Mean   :1.726   Mean   :1434   Mean   :  3.945       Mean   :0.5679  
 3rd Qu.:2.300   3rd Qu.:2000   3rd Qu.: 15.200       3rd Qu.:0.9300  
 Max.   :7.400   Max.   :2000   Max.   : 27.200       Max.   :3.5200  
    RA

In [15]:
# Save the dataset as `seoul_bike_sharing_converted.csv`
# write_csv(dataframe, "seoul_bike_sharing_converted.csv")
write_csv(seoul_bike_sharing, "seoul_bike_sharing_converted.csv")

# TASK: Normalize data


Columns `RENTED_BIKE_COUNT`, `TEMPERATURE`, `HUMIDITY`, `WIND_SPEED`, `VISIBILITY`, `DEW_POINT_TEMPERATURE`, `SOLAR_RADIATION`, `RAINFALL`, `SNOWFALL` are numerical variables/columns with different value units and range. Columns with large values may adversely influence (bias) the predictive models and degrade model accuracy. Thus, we need to perform normalization on these numeric columns to transfer them into a similar range.


In this project, you are asked to use Min-max normalization: 

**Min-max** rescales each value in a column by first subtracting the minimum value of the column from each value, and then divides the result by the difference between the maximum and minimum values of the column. So the column gets re-scaled such that the minimum becomes 0 and the maximum becomes 1.

$$x_{new} = \frac{x_{old} - x_{min}}{x_{max} - x_{min}}$$


_TODO:_ Apply min-max normalization on `RENTED_BIKE_COUNT`, `TEMPERATURE`, `HUMIDITY`, `WIND_SPEED`, `VISIBILITY`, `DEW_POINT_TEMPERATURE`, `SOLAR_RADIATION`, `RAINFALL`, `SNOWFALL`


In [16]:
# Use the `mutate()` function to apply min-max normalization on columns 
# `RENTED_BIKE_COUNT`, `TEMPERATURE`, `HUMIDITY`, `WIND_SPEED`, `VISIBILITY`, `DEW_POINT_TEMPERATURE`, `SOLAR_RADIATION`, `RAINFALL`, `SNOWFALL`
minmax_norm <- function(x){       
  (x-min(x))/(max(x)-min(x))}

bike_sharing_df %>%              
  mutate(TEMPERATURE = minmax_norm(TEMPERATURE),
         HUMIDITY = minmax_norm(HUMIDITY),
         WIND_SPEED = minmax_norm(WIND_SPEED),
         VISIBILITY = minmax_norm(VISIBILITY),
         DEW_POINT_TEMPERATURE = minmax_norm(DEW_POINT_TEMPERATURE),
         SOLAR_RADIATION = minmax_norm(SOLAR_RADIATION),
         RAINFALL = minmax_norm(RAINFALL),
         SNOWFALL = minmax_norm(SNOWFALL)) 
head(bike_sharing_df)

DATE       RENTED_BIKE_COUNT HOUR TEMPERATURE HUMIDITY  WIND_SPEED
1    01/12/2017 254                0   NA          0.3775510 0.29729730
2    01/12/2017 204                1   NA          0.3877551 0.10810811
3    01/12/2017 173                2   NA          0.3979592 0.13513514
4    01/12/2017 107                3   NA          0.4081633 0.12162162
5    01/12/2017  78                4   NA          0.3673469 0.31081081
6    01/12/2017 100                5   NA          0.3775510 0.20270270
7    01/12/2017 181                6   NA          0.3571429 0.17567568
8    01/12/2017 460                7   NA          0.3877551 0.12162162
9    01/12/2017 930                8   NA          0.3775510 0.14864865
10   01/12/2017 490                9   NA          0.2755102 0.06756757
11   01/12/2017 339               10   NA          0.2448980 0.16216216
12   01/12/2017 360               11   NA          0.2142857 0.17567568
13   01/12/2017 449               12   NA          0.2346939 0.18918919
14   01/12/2017 451               13   NA          0.2551020 0.21621622
15   01/12/2017 447               14   NA          0.2653061 0.27027027
16   01/12/2017 463               15   NA          0.3673469 0.43243243
17   01/12/2017 484               16   NA          0.5510204 0.56756757
18   01/12/2017 555               17   NA          0.5918367 0.21621622
19   01/12/2017 862               18   NA          0.6734694 0.18918919
20   01/12/2017 600               19   NA          0.7857143 0.22972973
21   01/12/2017 426               20   NA          0.8061224 0.20270270
22   01/12/2017 405               21   NA          0.8265306 0.10810811
23   01/12/2017 398               22   NA          0.8469388 0.20270270
24   01/12/2017 323               23   NA          0.8571429 0.13513514
25   02/12/2017 328                0   NA          0.8877551 0.14864865
26   02/12/2017 308                1   NA          0.8775510 0.08108108
27   02/12/2017 262                2   NA          0.8775510 0.20270270
28   02/12/2017 167                3   NA          0.8265306 0.29729730
29   02/12/2017  89                4   NA          0.8061224 0.27027027
30   02/12/2017  79                5   NA          0.8061224 0.20270270
⋮    ⋮          ⋮                 ⋮    ⋮           ⋮         ⋮         
8731 29/11/2018 1577              18   NA          0.4693878 0.12162162
8732 29/11/2018 1026              19   NA          0.4897959 0.25675676
8733 29/11/2018  818              20   NA          0.5510204 0.17567568
8734 29/11/2018  792              21   NA          0.5612245 0.28378378
8735 29/11/2018  726              22   NA          0.6020408 0.14864865
8736 29/11/2018  559              23   NA          0.5714286 0.25675676
8737 30/11/2018  474               0   NA          0.6326531 0.21621622
8738 30/11/2018  456               1   NA          0.6428571 0.27027027
8739 30/11/2018  300               2   NA          0.6734694 0.06756757
8740 30/11/2018  213               3   NA          0.7040816 0.12162162
8741 30/11/2018  116               4   NA          0.7244898 0.05405405
8742 30/11/2018  149               5   NA          0.6734694 0.06756757
8743 30/11/2018  293               6   NA          0.6938776 0.10810811
8744 30/11/2018  750               7   NA          0.7142857 0.10810811
8745 30/11/2018 1527               8   NA          0.6938776 0.14864865
8746 30/11/2018  809               9   NA          0.5816327 0.08108108
8747 30/11/2018  554              10   NA          0.5204082 0.10810811
8748 30/11/2018  642              11   NA          0.4387755 0.24324324
8749 30/11/2018  720              12   NA          0.3571429 0.17567568
8750 30/11/2018  740              13   NA          0.2448980 0.37837838
8751 30/11/2018  761              14   NA          0.2040816 0.29729730
8752 30/11/2018  768              15   NA          0.2040816 0.44594595
8753 30/11/2018  837              16   NA          0.2346939 0.20270270
8754 30/11/2018 1047              17   NA    

DATE       RENTED_BIKE_COUNT HOUR TEMPERATURE HUMIDITY WIND_SPEED VISIBILITY
1 01/12/2017 254               0    -5.2        37       2.2        2000      
2 01/12/2017 204               1    -5.5        38       0.8        2000      
3 01/12/2017 173               2    -6.0        39       1.0        2000      
4 01/12/2017 107               3    -6.2        40       0.9        2000      
5 01/12/2017  78               4    -6.0        36       2.3        2000      
6 01/12/2017 100               5    -6.4        37       1.5        2000      
  DEW_POINT_TEMPERATURE SOLAR_RADIATION RAINFALL SNOWFALL SEASONS HOLIDAY   
1 -17.6                 0               0        0        Winter  No Holiday
2 -17.6                 0               0        0        Winter  No Holiday
3 -17.7                 0               0        0        Winter  No Holiday
4 -17.6                 0               0        0        Winter  No Holiday
5 -18.6                 0               0        0        Winter  No Holiday
6 -18.7                 0               0        0        Winter  No Holiday
  FUNCTIONING_DAY
1 Yes            
2 Yes            
3 Yes            
4 Yes            
5 Yes            
6 Yes

In [17]:
# Print the summary of the dataset again to make sure the numeric columns range between 0 and 1
summary(bike_sharing_df)

     DATE           RENTED_BIKE_COUNT      HOUR        TEMPERATURE    
 Length:8760        Min.   :   2.0    Min.   : 0.00   Min.   :-17.80  
 Class :character   1st Qu.: 214.0    1st Qu.: 5.75   1st Qu.:  3.40  
 Mode  :character   Median : 542.0    Median :11.50   Median : 13.70  
                    Mean   : 729.2    Mean   :11.50   Mean   : 12.87  
                    3rd Qu.:1084.0    3rd Qu.:17.25   3rd Qu.: 22.50  
                    Max.   :3556.0    Max.   :23.00   Max.   : 39.40  
                    NA's   :295                       NA's   :11      
    HUMIDITY       WIND_SPEED      VISIBILITY   DEW_POINT_TEMPERATURE
 Min.   : 0.00   Min.   :0.000   Min.   :  27   Min.   :-30.600      
 1st Qu.:42.00   1st Qu.:0.900   1st Qu.: 940   1st Qu.: -4.700      
 Median :57.00   Median :1.500   Median :1698   Median :  5.100      
 Mean   :58.23   Mean   :1.725   Mean   :1437   Mean   :  4.074      
 3rd Qu.:74.00   3rd Qu.:2.300   3rd Qu.:2000   3rd Qu.: 14.800      
 Max.   :98.

In [18]:
# Save the dataset as `seoul_bike_sharing_converted_normalized.csv`
# write_csv(dataframe, "seoul_bike_sharing_converted_normalized.csv")
write_csv(bike_sharing_df, "seoul_bike_sharing_converted_normalized.csv")

## Standardize the column names again for the new datasets


Since you have added many new indicator variables, you need to standardize their column names again by using the following code:


In [19]:
# Dataset list
dataset_list <- c('seoul_bike_sharing.csv', 'seoul_bike_sharing_converted.csv', 'seoul_bike_sharing_converted_normalized.csv')

for (dataset_name in dataset_list){
    # Read dataset
    dataset <- read_csv(dataset_name)
    # Standardized its columns:
    # Convert all columns names to uppercase
    names(dataset) <- toupper(names(dataset))
    # Replace any white space separators by underscore, using str_replace_all function
    names(dataset) <- str_replace_all(names(dataset), " ", "_")
    # Save the dataset back
    write.csv(dataset, dataset_name, row.names=FALSE)
}

New names:
• `` -> `...1`
Rows: 8465 Columns: 15
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (4): DATE, SEASONS, HOLIDAY, FUNCTIONING_DAY
dbl (11): ...1, RENTED_BIKE_COUNT, HOUR, TEMPERATURE, HUMIDITY, WIND_SPEED, ...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 8465 Columns: 41
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (1): DATE
dbl (40): RENTED_BIKE_COUNT, TEMPERATURE, HUMIDITY, WIND_SPEED, VISIBILITY, ...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 8760 Columns: 14
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (4): DATE, SEASONS, HOLIDAY, FUNCTIONING_DAY
dbl (10): RENTED_BIKE_COUNT, HOU

# Next Steps


Great! Now that you have processed all of the necessary datasets, you are ready to perform exploratory data analysis to get some inital insights from them.


## Authors

<a href="https://www.linkedin.com/in/yan-luo-96288783/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkRP0321ENSkillsNetwork878-2022-01-01" target="_blank">Yan Luo</a>


### Other Contributors

Jeff Grossman


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description           |
| ----------------- | ------- | ---------- | ---------------------------- |
| 2021-04-08        | 1.0     | Yan        | Initial version created      |
|                   |         |            |                              |
|                   |         |            |                              |

## <h3 align="center"> © IBM Corporation 2021. All rights reserved. <h3/>
